Model family r1

y26s4824k264 opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've included any doc changes needed for <https://microsoft.github.io/autogen/>. See <https://github.com/microsoft/autogen/blob/main/CONTRIBUTING.md> to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

y26s4824k264 commented:
@microsoft-github-policy-service agree

y26s4824k264 commented:
@microsoft-github-policy-service agree






颜sk
***@***.***

ekzhu reviewed (changes requested):
I think the change can be simpler:

1. In create, when trying to parse the thinking tokens, we can check if the extra attribute "reasoning_content" is available in the choice, if it is available, then instead of parsing, we can just use the content directly for the thought field.

2. In create_stream. We shouldn't differentiate chunks from reasoning content or from main content. Chunks are only for UI purposes. Apply the same approach as in create to get the reasoning content at the end.

y26s4824k264 commented:
from openai import OpenAI import os # 初始化OpenAI客户端 client = OpenAI(  &nbsp; &nbsp;# 如果没有配置环境变量，请用百炼API Key替换：api_key="sk-xxx"  &nbsp; &nbsp;api_key = os.getenv("DASHSCOPE_API_KEY"),  &nbsp; &nbsp;base_url="https://dashscope.aliyuncs.com/compatible-mode/v1" ) reasoning_content = "" &nbsp;# 定义完整思考过程 answer_content = "" &nbsp; &nbsp; # 定义完整回复 is_answering = False &nbsp; # 判断是否结束思考过程并开始回复 # 创建聊天完成请求 completion = client.chat.completions.create(  &nbsp; &nbsp;model="deepseek-r1", &nbsp;# 此处以 deepseek-r1 为例，可按需更换模型名称  &nbsp; &nbsp;messages=[  &nbsp; &nbsp; &nbsp; &nbsp;{"role": "user", "content": "9.9和9.11谁大"}  &nbsp; &nbsp;],  &nbsp; &nbsp;stream=True,  &nbsp; &nbsp;# 解除以下注释会在最后一个chunk返回Token使用量  &nbsp; &nbsp;# stream_options={  &nbsp; &nbsp;# &nbsp; &nbsp; "include_usage": True  &nbsp; &nbsp;# } ) print("\n" + "=" * 20 + "思考过程" + "=" * 20 + "\n") for chunk in completion:  &nbsp; &nbsp;# 如果chunk.choices为空，则打印usage  &nbsp; &nbsp;if not chunk.choices:  &nbsp; &nbsp; &nbsp; &nbsp;print("\nUsage:")  &nbsp; &nbsp; &nbsp; &nbsp;print(chunk.usage)  &nbsp; &nbsp;else:  &nbsp; &nbsp; &nbsp; &nbsp;delta = chunk.choices[0].delta  &nbsp; &nbsp; &nbsp; &nbsp;# 打印思考过程  &nbsp; &nbsp; &nbsp; &nbsp;if hasattr(delta, 'reasoning_content') and delta.reasoning_content != None:  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;print(delta.reasoning_content, end='', flush=True)  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;reasoning_content += delta.reasoning_content  &nbsp; &nbsp; &nbsp; &nbsp;else:  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;# 开始回复  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;if delta.content != "" and is_answering == False:  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;print("\n" + "=" * 20 + "完整回复" + "=" * 20 + "\n")  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;is_answering = True  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;# 打印回复过程  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;print(delta.content, end='', flush=True)  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;answer_content += delta.content # print("=" * 20 + "完整思考过程" + "=" * 20 + "\n") # print(reasoning_content) # print("=" * 20 + "完整回复" + "=" * 20 + "\n") # print(answer_content)



我们不能过度聚焦于技术规范，更应该贴合落地在业务场景已经使用需求以及渲染的分开展示。且在 openai 的流式入口是两个字段表示思考与答案，我们也仅仅是在 Model.RI 这个模型判断之下进行如此使用，并未对原来的逻辑产生影响


**Implementation Approach:** &nbsp;
Focusing on *business-first* requirements rather than technical perfection, we: &nbsp;
1. Decouple rendering components from core logic &nbsp;
2. Align with OpenAI’s streaming protocol (`thought`/`answer` field structure) &nbsp;
3. Maintain Model.RI’s decision framework with zero legacy system impact &nbsp;




颜sk
***@***.***

ZakWork commented:
Doesn't look like it solves the problem. I have used your version but I am still getting the same error 
`UserWarning: Could not find <think>..</think> field in model response content. No thought was extracted.
  thought, content = parse_r1_content(content)`

ekzhu commented:
> 我们不能过度聚焦于技术规范，更应该贴合落地在业务场景已经使用需求以及渲染的分开展示。且在 openai 的流式入口是两个字段表示思考与答案，我们也仅仅是在 Model.RI 这个模型判断之下进行如此使用，并未对原来的逻辑产生影响

It's one thing to be practical but I think the current change is a bit too hacky. I am suggesting a first step that is simple. If we want to have more feature when it comes to presenting the thinking token to the consumer (e.g., UI), we can add a new field to `ModelStreamingChunkEvent` to indicate this is a reasoning content, rather than using `ThoughtEvent`, which is meant for a complete message.

ZakWork commented:
Hey, I have opened a new pull request #5961 to fix the following issue 
`UserWarning: Could not find <think>..</think> field in model response content. No thought was extracted. thought, content = parse_r1_content(content)`
I still get the issue of 
`openai.BadRequestError: Error code: 400 - {'error': {'message': 'deepseek-reasoner does not support successive user or assistant messages (messages[1] and messages[2] in your input). You should interleave the user/assistant messages in the message sequence.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_request_error'}}`
Any ideas where to start on this one?

y26s4824k264 commented:
> Hey, I have opened a new pull request #5961 to fix the following issue `UserWarning: Could not find <think>..</think> field in model response content. No thought was extracted. thought, content = parse_r1_content(content)` I still get the issue of `openai.BadRequestError: Error code: 400 - {'error': {'message': 'deepseek-reasoner does not support successive user or assistant messages (messages[1] and messages[2] in your input). You should interleave the user/assistant messages in the message sequence.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_request_error'}}` Any ideas where to start on this one?

![image](https://github.com/user-attachments/assets/b7c42fd9-a768-4576-8c82-7e2552fd8320)
I have removed the use of this function, and it seems that the error you encountered is not related to the part I modified.

y26s4824k264 commented:
> > 我们不能过度聚焦于技术规范，更应该贴合落地在业务场景已经使用需求以及渲染的分开展示。且在 openai 的流式入口是两个字段表示思考与答案，我们也仅仅是在 Model.RI 这个模型判断之下进行如此使用，并未对原来的逻辑产生影响
> 
> It's one thing to be practical but I think the current change is a bit too hacky. I am suggesting a first step that is simple. If we want to have more feature when it comes to presenting the thinking token to the consumer (e.g., UI), we can add a new field to `ModelStreamingChunkEvent` to indicate this is a reasoning content, rather than using `ThoughtEvent`, which is meant for a complete message.

Okay, I will add a "reasoning" field in ModelClientStreamingChunkEvent to represent the thinking content. I'll go and make the modification.

y26s4824k264 commented:
The modified ModelFamily.R1 supports the DeepSeek inference model deployed locally with ollama and is also compatible with the API services of cloud service online deployment.Make sure not to pull the wrong branch or code. I have removed the use of the parse_r1_content function.

ekzhu commented:
@y26s4824k264 thanks for your PR. 

The PR #5961 is almost at the finishing line. Perhaps instead of continuing with this PR, we can focus on the next step that is to make the UI layer understand which is thinking token and which isn't, once the other PR is merged?

y26s4824k264 commented:
![image](https://github.com/user-attachments/assets/93d07717-d153-43b5-b426-d56b21e66e22)

wow，God have mercy; I have provided the method for distinguishing the thinking process at the UI layer and unified the specifications.

> @y26s4824k264 thanks for your PR.
> 
> The PR #5961 is almost at the finishing line. Perhaps instead of continuing with this PR, we can focus on the next step that is to make the UI layer understand which is thinking token and which isn't, once the other PR is merged?

y26s4824k264 commented:
think token:  if message.reasoning  
content token: if message.content 

What remains is the matter of front-end rendering effects.

ekzhu commented:
Do you think maybe we should use `is_thought` flag to indicate the chunk is from reasoning? Right now, having two different content fields may create confusion which content to use from the point of view of the consumer of the event.

y26s4824k264 commented:
> Can you rebase this PR to the current main branch and add your changes from there.

ok

y26s4824k264 commented:
I have resubmitted a merge request from the main branch. I have run tests locally. The current merge branch request can be closed because I didn't run local tests when merging files and there were conflicts during online merging.

y26s4824k264 commented:
is #5989，
![image](https://github.com/user-attachments/assets/be198083-9204-4fbe-8cd3-1ab0c265215a)
